### ISSUE TITLE
copy assignment of outgoing iterators implicitely deleted

### ISSUE THREAD

**Issue by adl** [ticktac-project/tchecker#25]
If you change the outgoing loop of function `reachable()` in `reach-6.cc` from
```
    auto outgoing_range = builder.outgoing(node);
    for (auto it = outgoing_range.begin(); ! it.at_end(); ++it) {
```
to
```
    auto outgoing_range = builder.outgoing(node);
    auto it = outgoing_range.begin();
    for (it = outgoing_range.begin(); ! it.at_end(); ++it) {
```
you get a huge error message saying that
`tchecker::ts::builder_ok_t<TS, ALLOCATOR>::iterator_t<TS_ITERATOR>::operator=(const tchecker::ts::builder_ok_t<TS, ALLOCATOR>::iterator_t<TS_ITERATOR>&)` is `implicitly deleted because the default definition would be ill-formed`. 

Clang additionally notes that `copy assignment operator [...] is implicitly deleted because field '_it' has a deleted copy assignment operator`.

I find strange that the copy construction would work, but not the copy assignment.  

In my actual case those iterators is wrapped into a class that implement the successor iteration using Spot's interface, but that interface is "restartable": you may iterate over the successors more than once.  So I store two TChecker iterators: the current position, and a copy of `begin()`, so that I can assign it to the position (using the above method) to start another loop on request.

**Comment by fredher** [comment]
Copy construction works but not copy assignment because the class stores a reference. It has been changed to a pointer.

**Comment by adl** [comment]
My workaround for the missing copy assignment was to do:
```
    // instead of: start_ = start;
    start_.~ITERATOR();
    new(&start_)ITERATOR(start);
```
and it worked despite the reference.

However I've learned the hard way that those outgoing iterators hold a reference to the source state.  In my case the state is actually a `shared_state_ptr_t`, so the iterator is actually storing a reference to a shared pointer, and this gets a bit error-prone.  Consider a method like:
```
  builder_t::outgoing_iterator_t out(shared_state_ptr_t src)
  {
      return builder.outgoing(src).begin();
  };
```
Here the returned iterator has a dangling reference the local copy of `src`, and will cause issues when used.  Passing `src` by reference to `out` avoid the issue, but it took me a lot of time to figure out.   Would it be possible that when the state is a shared pointer, the iterator simply stores a copy?  This way the iterator will survive when passed around, even if the source state is not in scope anymore?

**Comment by fredher** [comment]
Solved in v0.3
